tests/net/connect_bind_send.rs

```diff
@@ -0,0 +1,559 @@
+    rustix::net::wsa_startup();
+
+    let localhost = IpAddr::V6(Ipv6Addr::LOCALHOST);
+    let addr = SocketAddr::new(localhost, 0);
+    let listener = rustix::net::socket(
+        AddressFamily::INET6,
+        SocketType::STREAM,
+        Protocol::default(),
+    )?;
+    rustix::net::bind(&listener, &addr).expect("bind");
+    rustix::net::listen(&listener, 1).expect("listen");
+
+    let local_addr = rustix::net::getsockname(&listener)?;
+    let local_addr = match local_addr {
+        SocketAddrAny::V6(v6) => SocketAddr::V6(v6),
+        other => panic!("unexpected socket address {:?}", other),
+    };
+    let sender = rustix::net::socket(
+        AddressFamily::INET6,
+        SocketType::STREAM,
+        Protocol::default(),
+    )?;
+    rustix::net::connect(&sender, &local_addr).expect("connect");
+    let request = b"Hello, World!!!";
+    let n = rustix::net::send(&sender, request, SendFlags::empty()).expect("send");
+    drop(sender);
+
+    // Not strictly required, but it makes the test simpler.
+    assert_eq!(n, request.len());
+
+    let accepted = rustix::net::accept(&listener).expect("accept");
+    let mut response = [0u8; 128];
+    let n = rustix::net::recv(&accepted, &mut response, RecvFlags::empty()).expect("recv");
+
+    // Not strictly required, but it makes the test simpler.
+    assert_eq!(n, request.len());
+
+    assert_eq!(request, &response[..n]);
+
+    #[cfg(windows)]
+    rustix::net::wsa_cleanup();
+
+    Ok(())
+}
+
+// Test `bind_any`.
+#[test]
+fn net_v4_bind_any() -> std::io::Result<()> {
+    #[cfg(windows)]
+    rustix::net::wsa_startup();
+
+    let localhost = Ipv4Addr::LOCALHOST;
+    let addr = SocketAddrAny::V4(SocketAddrV4::new(localhost, 0));
+    let listener =
+        rustix::net::socket(AddressFamily::INET, SocketType::STREAM, Protocol::default())?;
+    rustix::net::bind_any(&listener, &addr).expect("bind");
+    rustix::net::listen(&listener, 1).expect("listen");
+
+    let local_addr = rustix::net::getsockname(&listener)?;
+    let sender = rustix::net::socket(AddressFamily::INET, SocketType::STREAM, Protocol::default())?;
+    rustix::net::connect_any(&sender, &local_addr).expect("connect");
+    let request = b"Hello, World!!!";
+    let n = rustix::net::send(&sender, request, SendFlags::empty()).expect("send");
+    drop(sender);
+
+    // Not strictly required, but it makes the test simpler.
+    assert_eq!(n, request.len());
+
+    let accepted = rustix::net::accept(&listener).expect("accept");
+    let mut response = [0u8; 128];
+    let n = rustix::net::recv(&accepted, &mut response, RecvFlags::empty()).expect("recv");
+
+    // Not strictly required, but it makes the test simpler.
+    assert_eq!(n, request.len());
+
+    assert_eq!(request, &response[..n]);
+
+    #[cfg(windows)]
+    rustix::net::wsa_cleanup();
+
+    Ok(())
+}
+
+// Similar, but use V6.
+#[test]
+fn net_v6_bind_any() -> std::io::Result<()> {
+    #[cfg(windows)]
+    rustix::net::wsa_startup();
```

Comment:
A lot of code duplication in these tests; seems like this could at least be something like a `SocketTest` struct that could have e.g. `new_listen()` and its `impl Drop` could handle this bit.

Reply:
Good point. And it's also a problem that we're calling it for each test function, as test functions can run in parallel. I've now rewritten this to use a global constructor.